server/src/main/java/org/candlepin/model/AsyncJobStatus.java

```diff
@@ -101,6 +101,17 @@ public boolean isTerminal() {
         }
     }
 
+    /**
+     * A simple container object to collect all of the data to be stored in as a serialized field
+     */
+    @SuppressWarnings("checkstyle:VisibilityModifier")
+    private static class SerializedJobData {
+        public Map<String, String> metadata;
+        public JobArguments arguments;
+        public Object result;
```

Comment:
I would like to see this constrained this a bit more. There are a couple of options here:
1. Create a new abstract type (e.g. `AbstractJobResult`, then use `@JsonTypeInfo` and subtypes to allow some flexibility (e.g. `StringJobResult`, `ErrorJobResult`).
2. Create a single new concrete type `JobResult` and give it some fields that may or may not be present depending on the logical result of the job run (e.g. `stringData`, `exception`).

Reply:
I'm begrudgingly leaning toward option 1, but both start conflating the execute method of the job interface. I want to get away from the overbearing/non-standard input/output requirements we had imposed by Quartz.

I suppose it could be wrapped automatically by the job manager, but I'm not sure what we gain from doing so. As it is now with Object, it's as we discussed: objects which are properly annotated with \@JsonTypeInfo are deserialized properly. Everything else outside of the "standard" types Jackson natively handles will get converted to LinkedHashMaps.

Again, I'm willing to make this change, but I need to know what we gain here for the extra work/overhead of wrapping output from the job (or what we're losing by *not* doing this).

Reply:
It would be nice to have stronger typing. I'm not a fan of the interface being overly generic, as then the interface can be abused later on. It doesn't have to be resolved in this PR necessarily, and even if we do not revisit, that's okay, we can deal with any abuses of the interface with code review instead.

Reply:
What type of abuse are you predicting or worried about?

Reply:
> What type of abuse are you predicting or worried about?

Returning complex objects as job results, when the result should in most cases just be a primitive type. As I said previously though, I'm not too concerned; we can just keep an eye on it during subsequent PR review.

Reply:
Yeah. We have a few jobs which return complex objects, which is why I didn't want to restrict it straight away. I do want to keep this concern in the back of your mind, though. If we can come up with a way of cleaning this up without making it painful to write jobs, I'm open to exploring them and seeing where we land.